### ISSUE TITLE
CB API Caching Issue

### ISSUE THREAD

**Issue by momohossain** [carbonblack/cbapi-python#137]
I am unsure whether this has been fixed yet, but when I was working with the cbapi package a few months prior, I ran into an issue with the way the caching and the query objects are implemented.  

The cbapi package uses internal caching so it doesn't have to make multiple HTTP requests for the same information all the time.  The package also does not make an API request unless the code tries to _explicitly_ process the data.  For example, if I wanted to assign a policy onto a computer on CB Protection, it would not work unless I log an attribute from both the computer and policy object that I query for, or access the computer/policy object via the debugger, or print the computer/policy object to stdout, or something else in that respect. 

That being said, the CB Protection code I wrote worked when I tested it because all the information I was querying for was already stored onto my internal cache from me debugging multiple times.  It did not work when I pushed the code I wrote to production because the code was not written to attempt to retrieve the results from the queries it made, which resulted in no API requests, which in turn made the code attempt to work with `None` type values.  I could simply print everything I make a query for to stdout, but this could mean printing out fairly large chunks of data every time I run my code (which also would not take advantage of the internal caching).  

I was wondering if there was a way to have the query objects retrieve the results from the query in one go.

**Comment by jgarman** [comment]
Can you share your code? I definitely want to track down what you're seeing- as you mention, pulling data from the query result should be transparent. If you're changing data in an object, you have to call `.save()` on it to generate the POST or PUT request to update it on the server side.

**Comment by zacharyestep** [comment]
momohossain, Can we get more information? I'd love to be able to reproduce this and fix things if we can! I've been unable to reproduce - can you provide more information and maybe a stack trace? 

V/r,

**Comment by momohossain** [comment]
@jgarman I am just trying to access data via a GET request.  @zacharyestep When I encountered the error before, the stack trace stated that the code could not find the Computer or the Policy object. based on the parameters I gave the API.  The issue was I tried a script I wrote (that was supposed to change the policy of a computer) locally (where I wrote and tested all my code), and remotely (on a different box where we set up our integration) and the script that was run locally got results just fine, but the script run on the remote box could not find any data (it did not error out, it just did not get any results).  I can try to replicate my results again if I can somehow reset the cache for CB Protection on my local disk.
